src/rust/engine/dep_inference/src/javascript/mod.rs

```diff
@@ -18,42 +23,115 @@ include!(concat!(env!("OUT_DIR"), "/javascript/constants.rs"));
 include!(concat!(env!("OUT_DIR"), "/javascript/visitor.rs"));
 include!(concat!(env!("OUT_DIR"), "/javascript_impl_hash.rs"));
 
-#[derive(Serialize, Deserialize)]
+#[derive(Debug, Serialize, Deserialize)]
 pub struct ParsedJavascriptDependencies {
+    pub imports: HashMap<String, JavascriptImportInfo>,
+}
+
+#[derive(Debug, PartialEq, Serialize, Deserialize)]
+pub struct JavascriptImportInfo {
     pub file_imports: HashSet<String>,
     pub package_imports: HashSet<String>,
 }
 
+impl JavascriptImportInfo {
+    fn from_sets(file_imports: HashSet<String>, package_imports: HashSet<String>) -> Self {
+        JavascriptImportInfo {
+            file_imports,
+            package_imports,
+        }
+    }
+
+    #[cfg(test)]
+    fn new<F: AsRef<str>, T: AsRef<str>>(
+        file_imports: impl IntoIterator<Item = F>,
+        package_imports: impl IntoIterator<Item = T>,
+    ) -> Self {
+        JavascriptImportInfo::from_sets(
+            file_imports
+                .into_iter()
+                .map(|s| s.as_ref().to_string())
+                .collect(),
+            package_imports
+                .into_iter()
+                .map(|s| s.as_ref().to_string())
+                .collect(),
+        )
+    }
+}
+
+fn patterns_as_lookup(patterns: Vec<ImportPattern>) -> HashMap<String, Vec<String>> {
+    patterns
+        .into_iter()
+        .map(|pattern| (pattern.pattern, pattern.replacements))
```

Comment:
I see that this is just factoring out an existing behaviour, but for my background: do we care if we somehow end up with duplicate patterns here (in which case we'd be dropping input elements), or is that impossible by construction?

Reply:
Since the origin of these patterns are json objects the keys (pattern.pattern) are guaranteed to be unique.

I did an oopsie when I implemented this originally because I hadn't checked the protos spec in a while. Apparantly it does have [maps](https://protobuf.dev/programming-guides/proto2/#maps) nowadays.

Reply:
Ah, cool, no worries.